Investigate problem triggering assertion in Alignment#createAlignment()

*Before submission, please check that ...*

- [x] this is related specifically to recheck and not its extension (e.g. [recheck-web](https://github.com/retest/recheck-web)).
- [ ] ~the [documentation](https://docs.retest.de/) does not mention a fix (e.g. wrong API usage, ...).~
- [ ] ~the bug is not related to a customization (e.g. custom implementations).~
- [x] there are no open or closed issues that are related to this problem.

## Describe the Bug

When running the integration tests for `recheck-web`, an assertion in `recheck` is triggered. (_The problem is visible in the `recheck-web` integration tests, but stems from a problem in `recheck`_)

```
[ERROR]   RecheckRemoteWebElementIT.complete_html_should_be_same_as_driver:38 bestMatch and previousMatch have a match of 100%? At least paths should differ! html[1]/body[1]/form[3] # br # 1 == html[1]/body[1]/form[3] # br # 1
[ERROR]   RecheckRemoteWebElementIT.complete_html_should_be_same_as_web_element:49 bestMatch and previousMatch have a match of 100%? At least paths should differ! html[1]/body[1]/form[3] # br # 1 == html[1]/body[1]/form[3] # br # 1
```

This has not been present on earlier runs on the CI system when using Ubuntu 14.04.

### How to Reproduce?

Check out `retest-web`, run `mvn clean verify` on Fedora 31 or Ubuntu 18.04.

### Setup

- recheck: 1.8.0
- recheck-web: 8ef03a8eb49a136623f0d4608284e1eb28806e7a
- mvn 3.6.1 (Fedora 31), 
- OS: Fedora 31 / Ubuntu 18.04 (Bionic Beaver)
- Java version: Fedora 31: 13.0.1, Ubuntu: 11.0.2 (mvn compile) & 13.0.1 (rest of the steps)

### Steps to resolving this issue

This involves work across both projects.

- [ ] Reduce `form-page.html` to the minimal HTML needed to reproduce this issue (`recheck-web`)
- [ ] Create a test that can reproduce this issue (`recheck`)
- [ ] Fix underlying problem such that the test will not fail anymore  (`recheck`)
- [x] Removed `@Disabled` from the failing tests in `RecheckRemoteWebElementIT` (`recheck-web` )

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,7 @@ Table of Contents
 ### Bug Fixes
 
 * Add path info to filter, after they where missing in last release because of `FileSystemAlreadyExistsException` fix.
+* Fix some cases where applying deleted elements would not remove them from the Golden Master.
 
 ### New Features
 
@@ -48,7 +49,7 @@ Table of Contents
 ### Bug Fixes
 
 * Fix a rare `java.nio.file.FileSystemAlreadyExistsException` when accessing filters concurrently.
-* Fix 1-click-maintenance not respecting the key of the `AttributeDifference`. 
+* Fix 1-click-maintenance not respecting the key of the `AttributeDifference`.
 * Updated and now configurable URLs for rehub backend services
 
 ### New Features

--- a/src/main/java/de/retest/recheck/ui/descriptors/Element.java
+++ b/src/main/java/de/retest/recheck/ui/descriptors/Element.java
@@ -9,6 +9,7 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.ObjectUtils;
 import org.eclipse.persistence.oxm.annotations.XmlInverseReference;
@@ -129,15 +130,16 @@ public class Element implements Serializable, Comparable<Element> {
 	private List<Element> removeDeleted( final ActionChangeSet actionChangeSet,
 			final List<Element> oldContainedElements ) {
 		final Set<IdentifyingAttributes> deletedChanges = actionChangeSet.getDeletedChanges();
-		final List<Element> newContainedElements = new ArrayList<>( oldContainedElements.size() );
-
-		for ( final Element oldElement : oldContainedElements ) {
-			if ( !deletedChanges.contains( oldElement.getIdentifyingAttributes() ) ) {
-				newContainedElements.add( oldElement );
-			}
-		}
-
-		return newContainedElements;
+		// IdentifyingAttributes#equals cannot be used here, due to volatile attributes (e.g. outline)
+		final List<String> deletedIdentifiers = deletedChanges.stream() //
+				.map( IdentifyingAttributes::identifier ) //
+				.collect( Collectors.toList() );
+		return oldContainedElements.stream() //
+				.filter( element -> {
+					final IdentifyingAttributes attributes = element.getIdentifyingAttributes();
+					return !deletedIdentifiers.contains( attributes.identifier() );
+				} ) //
+				.collect( Collectors.toList() );
 	}
 
 	private List<Element> applyChangesToContainedElements( final ActionChangeSet actionChangeSet,

--- a/src/test/java/de/retest/recheck/ui/descriptors/ElementIT.java
+++ b/src/test/java/de/retest/recheck/ui/descriptors/ElementIT.java
@@ -1,0 +1,85 @@
+package de.retest.recheck.ui.descriptors;
+
+import static de.retest.recheck.ui.Path.fromString;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.awt.Rectangle;
+import java.util.Arrays;
+
+import org.junit.jupiter.api.Test;
+
+import de.retest.recheck.ui.review.ActionChangeSet;
+
+class ElementIT {
+
+	@Test
+	void applyChanges_should_remove_slightly_different_element() {
+		final Element form = new RootElement( "old", new IdentifyingAttributes( Arrays.asList( //
+				new PathAttribute( fromString( "html/body/form" ) ), //
+				new SuffixAttribute( 1 ), //
+				new TextAttribute( "type", "form" ) //
+		) ), new Attributes(), null, "", 0, "" );
+
+		final IdentifyingAttributes original = new IdentifyingAttributes( Arrays.asList( // 
+				new PathAttribute( fromString( "html/body/form/input" ) ), //
+				new SuffixAttribute( 1 ), //
+				new TextAttribute( "type", "input" ), //
+				OutlineAttribute.create( new Rectangle( 0, 0, 200, 100 ) ), //
+				OutlineAttribute.createAbsolute( new Rectangle( 100, 50, 200, 100 ) ) //
+		) );
+		final IdentifyingAttributes altered = new IdentifyingAttributes( Arrays.asList( // 
+				new PathAttribute( fromString( "html/body/form/input" ) ), //
+				new SuffixAttribute( 1 ), //
+				new TextAttribute( "type", "input" ), //
+				OutlineAttribute.create( new Rectangle( 0, 0, 250, 150 ) ), //
+				OutlineAttribute.createAbsolute( new Rectangle( 100, 20, 250, 150 ) ) //
+		) );
+
+		form.addChildren( new Element( "input", form, original, new Attributes(), null ) );
+
+		final ActionChangeSet changeSet = ActionChangeSetTestUtils.createEmptyActionChangeSet();
+		changeSet.addDeletedChange( altered );
+
+		final Element changed = form.applyChanges( changeSet );
+		assertThat( changed.getContainedElements() ).isEmpty();
+	}
+
+	@Test
+	void applyChanges_should_properly_handle_inserted_deleted_element() {
+		final Element form = new RootElement( "old", new IdentifyingAttributes( Arrays.asList( //
+				new PathAttribute( fromString( "html/body/form" ) ), //
+				new SuffixAttribute( 1 ), //
+				new TextAttribute( "type", "form" ) //
+		) ), new Attributes(), null, "", 0, "" );
+
+		final IdentifyingAttributes original = new IdentifyingAttributes( Arrays.asList( // 
+				new PathAttribute( fromString( "html/body/form/input" ) ), //
+				new SuffixAttribute( 1 ), //
+				new TextAttribute( "type", "input" ), //
+				OutlineAttribute.create( new Rectangle( 0, 0, 200, 100 ) ), //
+				OutlineAttribute.createAbsolute( new Rectangle( 100, 50, 200, 100 ) ) //
+		) );
+		final IdentifyingAttributes altered = new IdentifyingAttributes( Arrays.asList( // 
+				new PathAttribute( fromString( "html/body/form/input" ) ), //
+				new SuffixAttribute( 1 ), //
+				new TextAttribute( "type", "input" ), //
+				OutlineAttribute.create( new Rectangle( 0, 0, 250, 150 ) ), //
+				OutlineAttribute.createAbsolute( new Rectangle( 100, 20, 250, 150 ) ) //
+		) );
+
+		form.addChildren( new Element( "input", form, original, new Attributes(), null ) );
+
+		final ActionChangeSet changeSet = ActionChangeSetTestUtils.createEmptyActionChangeSet();
+		changeSet.addDeletedChange( original );
+		changeSet.addInsertChange( new Element( "input", form, altered, new Attributes(), null ) );
+
+		final Element changed = form.applyChanges( changeSet );
+
+		assertThat( changed.getContainedElements() ) //
+				.hasSize( 1 ) //
+				.element( 0 ) //
+				.satisfies( input -> {
+					assertThat( input.getIdentifyingAttributes() ).isEqualTo( altered );
+				} );
+	}
+}

